examples/xedge: Add Xedge example with BAS integration

JorgeGzm opened the pull request:
examples/xedge_demo: Add Xedge IoT Toolkit with BAS integration
    
    Add support for Xedge, an embedded software toolkit for IoT applications
    using Lua scripting with HTTP(S), WebSockets, MQTT, and device I/O.
    
    * netutils/xedge: Dependency manager that downloads BAS library
     and BAS-Resources, generates XedgeZip.c during build
    
    * examples/xedge_demo: Complete example showing Xedge integration
     with HTTP server, Lua runtime, and SNTP time synchronization

## Summary

I'm supporting @surfskidude to update xedge support with nuttx.

This commit covers conversations: 
https://github.com/apache/nuttx/pull/16352
https://github.com/apache/nuttx-apps/pull/3071

## Impact

``xedge_demo`` a lightweight Lua-based web framework for building secure, real-time IoT applications
=========================================================================

`Xedge <https://realtimelogic.com/products/xedge/>`_ is a robust IoT and web framework that is designed for microcontrollers. It is based on the industrial-grade Barracuda Application Server, designed for seamless OEM integration. Xedge accelerates embedded firmware development by providing a flexible, Lua-based environment and a full stack of industrial-strength protocols, including:

- OPC UA
- Modbus
- MQTT
- SMQ
- WebSockets
- HTTP/HTTPS

This Xedge port for NuttX comes pre-configured and requires:

- TCP/IP v4 and v6 support
- File System support
- 2 MB RAM allocated statically in ``xedge/xedge_main.c``

.. note::
   These instructions set up Xedge in **development mode**. Xedge supports many configuration options that differ between development and production builds. For production settings and optimization, refer to the general Xedge build instructions (details below).

Why Use Lua and Xedge in Embedded Systems
------------------------------------------

Great Lua developers don't treat it as a "better C"; they treat it as a complement. Lua is an extension language, which means it's designed to work alongside C, not replace it. Smart embedded programmers use C for performance-critical, low-level code and Lua for high-level business logic, such as processing sensor data and managing secure cloud connectivity.

Writing embedded business logic purely in C often means hundreds of lines of boilerplate code to manage memory, handle complex APIs, and handle errors. Lua, especially when paired with a framework like Xedge, lifts that burden. It provides high-level libraries and modules out of the box for protocols, networking, file systems, and more.

This shift doesn't just make development easier; it makes it faster. What used to take weeks in C can now be done in days. Lua's simplicity encourages rapid prototyping and quick iteration, which is essential in modern IoT and embedded development, where both time-to-market and security are critical. For a conceptual overview of why this hybrid development model is so powerful, check out the tutorial `Why Smart C Coders Love Lua.

## Testing

Run xedge example with qemu-armv8a board

```bash
 $ ./tools/configure.sh qemu-armv8a:xedge_demo
 $ make
```

Running with QEMU:

```bash
   $ qemu-system-aarch64 -cpu cortex-a53 -smp 4 -nographic \
     -machine virt,virtualization=on,gic-version=3 \
     -chardev stdio,id=con,mux=on -serial chardev:con \
     -global virtio-mmio.force-legacy=false \
     -netdev user,id=u1,hostfwd=tcp:127.0.0.1:8080-10.0.2.15:80,hostfwd=tcp:127.0.0.1:8443-10.0.2.15:443,hostfwd=tcp:127.0.0.1:10023-10.0.2.15:23 \
     -device virtio-net-device,netdev=u1,bus=virtio-mmio-bus.0 \
     -mon chardev=con,mode=readline -kernel ./nuttx
```

Running Xedge in NuttX terminal

```bash
   nsh> xedge_demo
    [    3.020000] [CPU0] Error: cannot set root to /mnt/lfs
    [    3.020000] [CPU0] Installing DiskIo failed!
    [    3.070000] [CPU0] Xedge: Server listening on IPv4 port 80
    [    3.080000] [CPU0] Xedge: SharkSSL server listening on IPv4 port 443
    [   13.720000] [CPU0] Xedge: Received invalid browser localStorage
    [   20.090000] [CPU1] 10.0.2.2 GET "rtl/assets/favicon.ico" t
    [   20.090000] [CPU1] Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138
    [   20.090000] [CPU1] .0.0.0 Safari/537.36
    [   20.090000] [CPU1] Host: 127.0.0.1:8080
    [   20.090000] [CPU1] Connection: keep-alive
    [   20.090000] [CPU1] sec-ch-ua-platform: "Linux"
    [   20.090000] [CPU1] User-Agent: a
    [   20.090000] [CPU1] Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/138
    [   20.090000] [CPU1] .0.0.0 Safari/537.36
    [   20.090000] [CPU1] sec-ch-ua: "Not)A;Brand";v="8", "Chromium";v="138", "Google Chrome";v="138"
    [   20.090000] [CPU1] sec-ch-ua-mobile: ?0
    [   20.090000] [CPU1] Accept: image/avif,image/webp,image/apng,image/svg+xml,image/*,*/*;q=0.8
    [   20.090000] [CPU1] Sec-Fetch-Site: same-origin
    [   20.090000] [CPU1] Sec-Fetch-Mode: no-cors
    [   20.090000] [CPU1] Sec-Fetch-Dest: image
    [   20.090000] [CPU1] Referer: http://127.0.0.1:8080/rtl/
    [   20.090000] [CPU1] Accept-Encoding: gzip, deflate, br, zstd
    [   20.090000] [CPU1] Accept-Language: en-US,en;q=0.9,pt-BR;q=0.8,pt;q=0.7
    [   20.090000] [CPU1] If-None-Match: 6866b8fe
    [   20.090000] [CPU1] If-Modified-Since: Thu, 03 Jul 2025 17:08:14 GMT
    [   20.090000] [CPU1]
    [   20.090000] [CPU0] 10.0.2.2 Response:
    T
    [   20.090000] [CPU0] Thu, 03 Jul 2025 19:32:12 GMT
    Content-Encoding: gzip
    Vary: Accept-Encoding
    Con
    [   20.090000] [CPU0] tent-Length: 3436
    Keep-Alive: Keep-Alive
```

Launch your web browser and access 127.0.0.1:8080

![image](https://github.com/user-attachments/assets/3910896d-2689-47e9-b4bb-8cffada1c093)

xiaoxiang781216 commented:
do you follow the suggestion from https://github.com/apache/nuttx-apps/pull/3071?

xiaoxiang781216 commented:
> ## Testing
> 
> This routine is incomplete and still needs to be fixed.
> 
> ```shell
> ./tools/configure.sh linum-stm32h753bi:netnsh
> ```
> 
> Open the menu config and enable the flags bellow:
> 
> NETUTILS_NTPCLIENT=y ALLOW_GPL_COMPONENTS=y EXAMPLES_XEDGE=y
> 
> ```shell
> $ make -j
> Create version.h
> LN: platform/board to /home/jaga/nuttxspace/apps/platform/dummy
> Downloading BAS from hash 9f74a2f778b002ad8441471b8a7a5b13172dbe76...
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> 100 3867k    0 3867k    0     0  4078k      0 --:--:-- --:--:-- --:--:-- 4078k
> Downloading BAS-Resources from hash 227a4b998300fa4cfde871dc7dac92c09e1636c2...
>   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
>                                  Dload  Upload   Total   Spent    Left  Speed
>   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
> 100  813k    0  813k    0     0  1192k      0 --:--:-- --:--:-- --:--:-- 1192k
> Creating XedgeZip.c
> Register: nsh
> Register: sh
> Register: ping
> Register: dd
> Register: telnetd
> Register: xedge
> Register: alarm
> CPP:  /home/jaga/nuttxspace/nuttx/boards/arm/stm32h7/linum-stm32h753bi/scripts/flash.ld-> /home/jaga/nuttxspace/nuttx/boards/arm/stm32h7LD: nuttx
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/nuttx section `.bss' will not fit in region `sram'
> arm-none-eabi-ld: region `sram' overflowed by 1621920 bytes
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `luaD_throw':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:11775: undefined reference to `longjmp'
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `luaD_rawrunprotected':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:11801: undefined reference to `setjmp'
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `lua_strx2number':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:16759: undefined reference to `localeconv'
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `l_str2d':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:16857: undefined reference to `localeconv'
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `tostringbuff':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:16956: undefined reference to `localeconv'
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `num2straux':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:21029: undefined reference to `localeconv'
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `quotefloat':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:21149: undefined reference to `localeconv'
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `l_strcmp':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:23289: undefined reference to `strcoll'
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `os_execute':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:25147: undefined reference to `system'
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `os_setlocale':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:25423: undefined reference to `setlocale'
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `read_number':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:26782: undefined reference to `localeconv'
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `DeadProp_save':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:83645: undefined reference to `setjmp'
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `LockNode_setOwner':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:85492: undefined reference to `setjmp'
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `cachelevel':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:86369: undefined reference to `setjmp'
> arm-none-eabi-ld: /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:86478: undefined reference to `setjmp'
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `resetstate':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:86702: undefined reference to `setjmp'
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o):/home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:86798: more undefined references to `setjmp' follow
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `SlException_set':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:92496: undefined reference to `longjmp'
> arm-none-eabi-ld: /home/jaga/nuttxspace/nuttx/staging/libapps.a(BAS.c.home.jaga.nuttxspace.apps.examples.xedge_1.o): in function `SXmlRoot_parse':
> /home/jaga/nuttxspace/apps/examples/xedge/BAS/src/BAS.c:92507: undefined reference to `setjmp'
> make[1]: *** [Makefile:217: nuttx] Error 1
> make: *** [tools/Unix.mk:552: nuttx] Error 2
> ```

why still contain error in your log?

JorgeGzm commented:
> do you follow the suggestion from #3071?

Yes, I tried to implement all the suggestions, please if I forget something that is violating the nuttx rules tell me.

JorgeGzm commented:
> > ## Testing
> > This routine is incomplete and still needs to be fixed.

I'm trying to talk to @surfskidude to fix the build error.

xiaoxiang781216 commented:
> > do you follow the suggestion from #3071?
> 
> Yes, I tried to implement all the suggestions, please if I forget something that is violating the nuttx rules tell me.

examples and 3rd party project shouldn't bundle, but split into the different patch.

surfskidude commented:
>> why still contain error in your log?

The two functions longjmp and setjmp, which are part of the standard C library, are required by Lua. The fact that you get a link error for setjmp alone indicates that there is an issue with the NuttX stm32h753bi build. Since NuttX is POSIX compliant, it should also include function setlocale, but this function can be removed if it’s generally not part of NuttX. However, that would mean NuttX is not POSIX-compliant. Regarding the SRAM issue: Note that Xedge is not designed for microcontrollers without external RAM, so your RAM sections must be adjusted accordingly. I have configured a 2Mbyte allocator in xedge_main.c. This is a static array and can typically not be in sram on a micro.

Please note that I only tested Xedge in the host simulator.

acassis commented:
> > > why still contain error in your log?
> 
> The two functions longjmp and setjmp, which are part of the standard C library, are required by Lua. The fact that you get a link error for setjmp alone indicates that there is an issue with the NuttX stm32h753bi build. Since NuttX is POSIX compliant, it should also include function setlocale, but this function can be removed if it’s generally not part of NuttX. However, that would mean NuttX is not POSIX-compliant. Regarding the SRAM issue: Note that Xedge is not designed for microcontrollers without external RAM, so your RAM sections must be adjusted accordingly. I have configured a 2Mbyte allocator in xedge_main.c. This is a static array and can typically not be in sram on a micro.
> 
> Please note that I only tested Xedge in the host simulator.

I think it is happening because you forgot to enable CONFIG_ARCH_SETJMP_H

cederom reviewed (changes requested):
Thank you @JorgeGzm :-)

This seems to be a copy of https://github.com/apache/nuttx-apps/pull/3071 but why it was closed before problem resolution and merge? I guess you took over the work right? :-)

Please take a look at:
1. https://github.com/apache/nuttx/blob/master/CONTRIBUTING.md
2. https://nuttx.apache.org/docs/latest/contributing/index.html
3. https://nuttx.apache.org/docs/latest/

Hopefully we can make it this time when fixes are ready and all works as expected, thanks again :-)

JorgeGzm commented:
@xiaoxiang781216 Could you check if there's anything else that needs to be fixed.

In the PR https://github.com/apache/nuttx/pull/16665 I added an example of how to use xedge with qemu-armv8a board.

cederom commented:
Hmm I have this build error on FreeBSD, any hints welcome :-)

```
% ./tools/configure.sh -B qemu-armv8a:xedge
% gmake -j24 CROSSDEV=aarch64-none-elf-
(..)
done
Register: iperf
Register: xedge
Register: nxplayer
Register: renew
Register: hello
Register: dd
Register: nsh
Register: sh
Register: nxrecorder
Register: ping
Register: fdtdump
Register: ostest
Register: getprime
Register: taskset
Register: telnetd
Register: smp
LD: nuttx
aarch64-none-elf-ld: /usr/lib/libm.a(e_acos.o): Relocations in generic ELF (EM: 62)
aarch64-none-elf-ld: /usr/lib/libm.a(e_acos.o): Relocations in generic ELF (EM: 62)
aarch64-none-elf-ld: /usr/lib/libm.a(e_acos.o): Relocations in generic ELF (EM: 62)
aarch64-none-elf-ld: /usr/lib/libm.a(e_acos.o): Relocations in generic ELF (EM: 62)
aarch64-none-elf-ld: /usr/lib/libm.a: error adding symbols: file in wrong format
gmake[1]: *** [Makefile:190: nuttx] Error 1
gmake: *** [tools/Unix.mk:552: nuttx] Error 2

% aarch64-none-elf-gcc --version
aarch64-none-elf-gcc (FreeBSD Ports Collection for aarch64noneelf) 11.3.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

% uname -a
FreeBSD hexagon 14.2-RELEASE-p1 FreeBSD 14.2-RELEASE-p1 GENERIC amd64
```

cederom commented:
Yup :-) GCC 14.2 AARCH64 needs to be used :-)

```
% ./tools/configure.sh -B qemu-armv8a:xedge
% gmake -j24 CROSSDEV=aarch64-none-elf-
(..)
Register: ping
Register: renew
Register: iperf
Register: dd
Register: hello
Register: xedge
Register: fdtdump
Register: nsh
Register: sh
Register: nxrecorder
Register: nxplayer
Register: getprime
Register: taskset
Register: telnetd
Register: ostest
Register: smp
LD: nuttx
Memory region         Used Size  Region Size  %age Used
CP: nuttx.hex
CP: nuttx.bin

% aarch64-none-elf-gcc --version
aarch64-none-elf-gcc (gcc) 14.2.1 20241119
Copyright (C) 2024 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


% qemu-system-aarch64
hexagon%    $ qemu-system-aarch64 -cpu cortex-a53 -smp 4 -nographic \
     -machine virt,virtualization=on,gic-version=3 \
     -chardev stdio,id=con,mux=on -serial chardev:con \
     -global virtio-mmio.force-legacy=false \
     -netdev user,id=u1,hostfwd=tcp:127.0.0.1:8080-10.0.2.15:80,hostfwd=tcp:127.0.0.1:8443-10.0.2.15:443,hostfwd=tcp:127.0.0.1:10023-10.0.2.15:23 \
     -device virtio-net-device,netdev=u1,bus=virtio-mmio-bus.0 \
     -mon chardev=con,mode=readline -kernel ./nuttx
zsh: command not found: $
hexagon% qemu-system-aarch64 -cpu cortex-a53 -smp 4 -nographic \
     -machine virt,virtualization=on,gic-version=3 \
     -chardev stdio,id=con,mux=on -serial chardev:con \
     -global virtio-mmio.force-legacy=false \
     -netdev user,id=u1,hostfwd=tcp:127.0.0.1:8080-10.0.2.15:80,hostfwd=tcp:127.0.0.1:8443-10.0.2.15:443,hostfwd=tcp:127.0.0.1:10023-10.0.2.15:23 \
     -device virtio-net-device,netdev=u1,bus=virtio-mmio-bus.0 \
     -mon chardev=con,mode=readline -kernel ./nuttx
telnetd [8:100]

NuttShell (NSH) NuttX-10.4.0
nsh> uname -a
NuttX  10.4.0 ffecbaa4a2 Jul  4 2025 16:02:01 arm64 qemu-armv8a
nsh> help
help usage:  help [-v] [<cmd>]

    .           cmp         free        mkdir       pwd         umount
    [           dirname     get         mkfatfs     rm          unset
    ?           df          help        mkrd        rmdir       uptime
    alias       dmesg       hexdump     mount       set         usleep
    unalias     echo        ifconfig    mv          sleep       watch
    arp         env         ifdown      nfsmount    source      wget
    basename    exec        ifup        nslookup    test        xd
    break       exit        kill        pidof       time        wait
    cat         expr        pkill       printf      true
    cd          false       ls          ps          truncate
    cp          fdinfo      md5         put         uname

Builtin Apps:
    dd            iperf         ostest        smp
    fdtdump       nsh           ping          taskset
    getprime      nxplayer      renew         telnetd
    hello         nxrecorder    sh            xedge
nsh> free
      total       used       free    maxused    maxfree  nused  nfree name
  126701568     142848  126558720     143544  126558064     51      3 Umem
nsh> xedge
[   34.940000] [CPU0] Error: cannot set root to /mnt/lfs
[   34.940000] [CPU0] Installing DiskIo failed!
[   35.000000] [CPU0] Xedge: Server listening on IPv4 port 80
[   35.000000] [CPU0] Xedge: SharkSSL server listening on IPv4 port 443
```

JorgeGzm commented:
I followed the qemu-armv8a board tutorial to set up the build environment:

https://nuttx.apache.org/docs/latest/platforms/arm64/qemu/boards/qemu-armv8a/index.html

# Compile Toolchain
Host environment: GNU/Linux: Ubuntu 18.04 or greater

Download and Install

```bash
wget https://developer.arm.com/-/media/Files/downloads/gnu/11.2-2022.02/binrel/gcc-arm-11.2-2022.02-x86_64-aarch64-none-elf.tar.xz
xz -d gcc-arm-11.2-2022.02-x86_64-aarch64-none-elf.tar.xz
tar xf gcc-arm-11.2-2022.02-x86_64-aarch64-none-elf.tar
```

Put gcc-arm-11.2-2022.02-x86_64-aarch64-none-elf/bin/ on your host PATH environment variable, like:

```bash
export PATH=$PATH:/opt/software/arm/linaro-toolchain/gcc-arm-11.2-2022.02-x86_64-aarch64-none-elf/bin
```
check the toolchain:

```bash
aarch64-none-elf-gcc -v
```

acassis commented:
@xiaoxiang781216 @cederom PTAL

cederom reviewed (approved):
Thank you @JorgeGzm amazing work! :-)

cederom commented:
> Why does GitHub say it cannot be merged? Need to rebase or something?

All comments must be resolved :-) There is still one open conversation :-)

JorgeGzm commented:
> > Por que o GitHub diz que não é possível mesclar? Precisa fazer rebase ou algo assim?
> 
> Todos os comentários devem ser resolvidos :-) Ainda há uma conversa em aberto :-)

I believe I have resolved all the improvements, can you please tell me what part is missing.

JorgeGzm commented:
> > Why does GitHub say it cannot be merged? Need to rebase or something?
> 
> All comments must be resolved :-) There is still one open conversation :-)

xiaoxiang781216 reviewed (changes requested):
Before merged, this patch should separate xedge to independent project and split one patch to two: one is xedge project and another is xedge demo.

xiaoxiang781216 commented:
> > Before merged, this patch should separate xedge to independent project and split one patch to two: one is xedge project and another is xedge demo.
> 
> Thank you for the feedback. I understand now that the project should be split into two separate patches:
> 
>     * Xedge Library - The core toolkit as an independent, reusable component in **apps/xxx/xedge/**
> 
>     * Xedge Example - A simple demonstration of library usage in **apps/examples/xedgedemo/**
> 
> 
> Before proceeding with the implementation, I have a question about the proper location for the Xedge library. Would **apps/netutils/xedge/** be the appropriate directory, or would you recommend a different location such as **apps/framework/xedge/** ?
> 

netutils or system is a good candidate. @cederom what's your opinion?

JorgeGzm commented:
@xiaoxiang781216 @cederom

I have split the project into two separate paths and updated all the documentation accordingly.

However, I have encountered a new issue: when I run **make distclean** or **make app_distclean**, the downloaded folders BAS and BAS-Resources are not being deleted.

The problem appears to be in the distclean:: routine in the **apps/netutils/xedge/Makefile** file. I need help to resolve this issue.

xiaoxiang781216 commented:
> @xiaoxiang781216 @cederom
> 
> I have split the project into two separate paths and updated all the documentation accordingly.
> 
> However, I have encountered a new issue: when I run **make distclean** or **make app_distclean**, the downloaded folders BAS and BAS-Resources are not being deleted.
> 
> The problem appears to be in the distclean:: routine in the **apps/netutils/xedge/Makefile** file. I need help to resolve this issue.

where your patch? there are many demo Makefile which download, build and clean the package correctly, you just follow one of demo carefully.

xiaoxiang781216 commented:
@JorgeGzm why do you add me review again before you fix the issue I mention?

acassis commented:
@xiaoxiang781216 let's merge the example with the BAS integrated and then he will separate it, otherwise this PR never will completed...

acassis commented:
> but why not done in the correct way? it's a simple modification.

@xiaoxiang781216 I think we missed it, he already separated xedge server and xedge application

JorgeGzm commented:
> but why not done in the correct way? it's a simple modification.

Sorry for the confusion. I was working on two different computers and only just realized that I had been working on diferent branches. I’ve now submitted the update, and I believe the issue has been fully resolved.

JorgeGzm commented on examples/xedge_demo/xedge_main.c:
```diff
@@ -0,0 +1,278 @@
+ * https://realtimelogic.com/ba/doc/en/C/introduction.html#oo_c
+ */
+
+#define ltmgr ltMgr
```
if I remove the defines, the application doesn't compile. What exactly is your suggestion?

xiaoxiang781216 replied:
The application is still compiled, but nxstyle will report the "Mixed Case" warning, but it isn't a fatal error since the 3rd party library may use the camel case coding style.

JorgeGzm replied:
I removed all the #define aliases that were used to hide camelCase names from the Barracuda App Server library.

The code now uses the original names (e.g., ThreadMutex_set, LThreadMgr_run) directly.
The application compiles and runs fine with this change.

Let me know if anything else is needed.

acassis replied:
@JorgeGzm these CameCase symbols need to be included to g_white_prefix[] inside nuttx/tools/nxstyle.c to avoid compilation errors, please open a PR there to include them.

@xiaoxiang781216 @simbit18 maybe the whitelist should become a separated file instead of hardcoded inside nxstyle.c (and also to avoid it growing too much there)

JorgeGzm commented:
I’ve updated the whitelist in PR [#16665](https://github.com/apache/nuttx/pull/16665) to allow mixed-case identifiers used by the Xedge (Barracuda App Server) example.

The following prefixes were added to g_white_prefix[]:
- Ba
- Thread
- LThread
- Http
- Disk
- Xedge
    
The following exact identifiers were added to g_white_content_list[]:
- baGetUnixTime
- baParseDate
- setDispExit
- xedgeInitDiskIo
- xedgeOpenAUX
- platformInitDiskIo
- ltMgr
- Lt

JorgeGzm commented:
Talking to @acassis, he asked to submit a separate commit changing the whitelist that is in PR https://github.com/apache/nuttx/pull/16759

xiaoxiang781216 commented on examples/xedge_demo/xedge_main.c:
```diff
@@ -0,0 +1,254 @@
+ *   Copyright (C) 2025. All rights reserved.
+ *   Author: Real Time Logic
+ *
+ * Redistribution and use in source and binary forms, with or without
```
need apache copyright

xiaoxiang781216 commented on examples/xedge_demo/Make.defs:
```diff
@@ -0,0 +1,24 @@
+
+ifneq ($(CONFIG_EXAMPLES_XEDGE_DEMO),)
+CONFIGURED_APPS += $(APPDIR)/examples/xedge_demo
+CFLAGS += ${INCDIR_PREFIX}$(APPDIR)/netutils/xedge/BAS/inc ${INCDIR_PREFIX}$(APPDIR)/netutils/xedge/BAS/inc/arch/Posix ${INCDIR_PREFIX}$(APPDIR)/netutils/xedge/BAS/inc/arch/NET/Posix -DNO_INIT_DISK_IO -DUSE_DBGMON=1 -DUSE_IPV6
```
remove CFLAGS

JorgeGzm replied:
Ok, I changed CFLAGS to netutils/xedge/Make.defs, using the lvgl package as a reference

JorgeGzm commented on netutils/xedge/Make.defs:
```diff
@@ -0,0 +1,35 @@
+
+CONFIGURED_APPS += $(APPDIR)/netutils/xedge
+
+CFLAGS += ${INCDIR_PREFIX}$(APPDIR)/netutils/xedge/BAS/inc
```
Ok, I removed the duplicate routine in the Makefile, if I remove this routine in Make.defs the compilation breaks.

```
-CFLAGS += ${INCDIR_PREFIX}$(CURDIR)/$(BAS_UNPACKNAME)/inc
-CFLAGS += ${INCDIR_PREFIX}$(CURDIR)/$(BAS_UNPACKNAME)/inc/arch/Posix
-CFLAGS += ${INCDIR_PREFIX}$(CURDIR)/$(BAS_UNPACKNAME)/inc/arch/NET/Posix
-CFLAGS += -DNO_INIT_DISK_IO -DUSE_DBGMON=1 -DUSE_IPV6
```

What other suggestion?
